man/rec2csv.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.10.
.TH REC2CSV "1" "April 2014" "rec2csv 1.7" "User Commands"
.SH NAME
rec2csv \- rec to csv converter
.SH SYNOPSIS
.B rec2csv
[\fIOPTIONS\fR]... [\fIREC_FILE\fR]
.SH DESCRIPTION
Convert rec data into csv data.
.TP
\fB\-d\fR, \fB\-\-delim\fR=\fIchar\fR
sets the deliminator (default ',')
.TP
\fB\-t\fR, \fB\-\-type\fR=\fITYPE\fR
record set to convert to csv; if this parameter
is omitted then the default record set is used
.TP
\fB\-S\fR, \fB\-\-sort\fR=\fIFIELDS\fR
sort the output by the specified fields.
.TP
\fB\-\-help\fR
print a help message and exit.
.TP
\fB\-\-version\fR
show version and exit.
.SH AUTHOR
Written by Jose E. Marchesi.
.SH "REPORTING BUGS"
Report bugs to: bug\-recutils@gnu.org
.br
GNU recutils home page: <http://www.gnu.org/software/recutils/>
.br
General help using GNU software: <http://www.gnu.org/gethelp/>
.SH COPYRIGHT
Copyright \(co 2010, 2011, 2012, 2013, 2014 Jose E. Marchesi.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B rec2csv
is maintained as a Texinfo manual.  If the
.B info
and
.B rec2csv
programs are properly installed at your site, the command
.IP
.B info recutils
.PP
should give you access to the complete manual.


NEWS
This NEWS file records noteworthy changes, very tersely.
See the manual for detailed information.

  Copyright 2010-2014 Jose E. Marchesi.

  Copying and distribution of this file, with or without modification,
  are permitted in any medium without royalty provided the copyright
  notice and this notice are preserved.

 Version 1.7 (25 March 2014)
* Utilities:
  . rec2csv now supports a -d option to change the delimiter
    character in CSV files.
  . It is now possible to specify several recfiles in the
    command line containing anonymous records.
* Format:
  . Support for restricting the fields that can appear in records
    with the new %allowed special field.
  . If a tab separates a field name and a field value is now
    ignored and not considered part of the field value.
* librec:
  . The current version of the library is now 1.
* Emacs mode:
  . It now possible to jump several records while navigating
    using a prefix argument to the next/previous commands.
* Documentation:
  . The examples in the manual have been improved, and well as many
    other aspects: formatting, English, etc.  Special thanks to Karl
    Berry for pointing out the problems and suggesting solutions.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.6 (03 November 2013)
* Utilities:
  . Support for joins and foreign keys.
  . Grouping records by one or more fields.
  . Sorting recors by several fields.
  . Specifying rewrite rules in field expressions.
  . Writing field values in rows.
  . Dot-notation in selection expressions.
  . Support for the >= and <= operators in selection expressions.
  . Aggregated functions: sum, min, max, avg.
  . New bash loadable builtin `readrec'.
* Format:
  . New UUID field type.
  . User provided arbitrary constraints with %constraint.
  . Pre-defined constants MIN and MAX can now be used in ranges.
* librec:
  . New high-level API: rec_db_query, rec_db_insert, rec_db_delete and
    rec_db_set.
  . Documentation in the rec.h header file extended and improved.
* Emacs mode:
  . Texinfo manual describing the usage of rec-mode.
  . Support for a mode hook.
  . Comments are now supported in records.
  . add-change-log-entry in a recfile now does the right thing.
  . Commands for copying/killing fields and records.
  . Support for undo in navigation mode.
  . Field folding.
  . Editing methods for the several field types.
  . Statistics.
* Documentation:
  . The user manual has been restructured and expanded.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.5 (13 January 2012)
* The utilities will now ask interactively for a password if it was
  not provided with the -s command line option.  This avoids security
  problems related to shell history files.
* Support for octal and hexadecimal numbers has been added.  They can
  be used in both the records and selection expressions.
* It is now possible to select a given number of random records in
  many of the utilities using the -m command line option.
* The -n option now accepts a list of indexes, supporting ranges.
* The new -U (uniq) option for recsel removes duplicated fields in the
  output records.
* The new -q (quick) option allows to quickly search for the desired
  record without having to provide a complete selection expression.
* Auto generated fields are now considered integers by default.  This
  avoids repetitive patterns in record descriptors involving %auto and
  %type.
* Tab characters are now allowed in blank lines betwwen records.
* The API in rec.h is now better documented with comments, and
  improved.
* recfix now exits with an error status if there is a parse error in
  some input file.
* The usage of the internal data structures has been _vastly_
  improved, resulting in a much faster operation.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.4 (03 November 2011)
* Support for encryption with the new %confidential special field.
* Support for named types with the new %typedef special field.
* Support for sorting with the new %sort special field.
* Recutils support for org-babel is now included in the distribution
  as ob-rec.el.  It allows to read rec data into org documents in both
  raw and table formats.
* It is now possible to use double quotes for delimiting string
  literals in selection expressions, as well as single quotes.
  Unescapted newline characters are also supported.
* The manual has been vastly improved.
* recfix now supports several operations: check, sort, encrypt and
  decrypt.
* recins now generates parseable locale independent dates when
  generating time-stamps in auto-fields.
* Lots of bug fixes.

 Version 1.3 (11 February 2011)
* A quite bad (and embarrassing) bug introduced in 1.2 has been
  fixed.  The bug was causing recins, recdel and recset to report
  random integrity failures, due to a uninitialized variable.
* New special field %size, that allows to define constraints in the
  size of record sets.
* Several fixes in the user manual.
* Little portability issue that prevented one of the tests to compile
  in macos x has been fixed.
* Translations updates.

 Version 1.2 (4 February 2011)
* New utility rec2csv, to export rec data to csv files.
* Support for auto-incremented fields and timestamps with the new
  %auto field in record descriptors.
* Support for comments in enum type descriptions.
* recset now supports a new "set or add field" operation, implementing
  the common case of setting the value of a field in case it exists,
  or else add it to the record.
* The flavor of regular expressions accepted in both selection
  expressions and type descriptions is now documented in the user
  manual.
* Translations for German, Finnish, Swedish and Dutch.
* recdel now performs an integrity check before to complete the
  deletion.
* The utilities no longer "expand" the contents of remote descriptors
  in the local ones.

utils/rec2csv.c
/* -*- mode: C -*-
 *
 *       File:         rec2csv.c
 *       Date:         Mon Jan 31 22:12:29 2011
 *
 *       GNU recutils - rec to csv converter.
 *
 */

/* Copyright (C) 2011-2014 Jose E. Marchesi */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>

#include <getopt.h>
#include <string.h>
#include <stdlib.h>
#include <xalloc.h>
#include <gettext.h>
#define _(str) gettext (str)

#include <csv.h>
#include <rec.h>
#include <recutl.h>

/* Forward declarations.  */
static void rec2csv_parse_args (int argc, char **argv);
static bool rec2csv_process_data (rec_db_t db);
static rec_fex_t rec2csv_determine_fields (rec_rset_t rset);
static void rec2csv_generate_csv (rec_rset_t rset, rec_fex_t fex);

/*
 * Types
 */

/*
 * Global variables
 */

char             *rec2csv_record_type    = NULL;
rec_fex_t         rec2csv_sort_by_fields = NULL;
char              rec2csv_delim          = ',';

/*
 * Command line options management
 */

enum
  {
    COMMON_ARGS,
    RECORD_TYPE_ARG,
    SORT_ARG
  };

static const struct option GNU_longOptions[] =
  {
    COMMON_LONG_ARGS,
    {"type", required_argument, NULL, RECORD_TYPE_ARG},
    {"sort", required_argument, NULL, SORT_ARG},
    {NULL, 0, NULL, 0}
  };


/*
 * Functions.
 */

void
recutl_print_help (void)
{
  /* TRANSLATORS: --help output, rec2csv synopsis.
     no-wrap */
  printf (_("\
Usage: rec2csv [OPTIONS]... [REC_FILE]\n"));

  /* TRANSLATORS: --help output, rec2csv short description.
     no-wrap */
  fputs (_("\
Convert rec data into csv data.\n"), stdout);

  puts ("");
  /* TRANSLATORS: --help output, rec2csv options.
     no-wrap */
  fputs (_("\
  -d, --delim=char                    sets the deliminator (default ',')\n\
  -t, --type=TYPE                     record set to convert to csv; if this parameter\n\
                                        is omitted then the default record set is used\n\
  -S, --sort=FIELDS                   sort the output by the specified fields.\n"),
         stdout);

  recutl_print_help_common ();
  puts ("");
  recutl_print_help_footer ();
}

static void
rec2csv_parse_args (int argc,
                    char **argv)
{
  int ret;
  char c;

  while ((ret = getopt_long (argc,
                             argv,
                             "t:S:d:",
                             GNU_longOptions,
                             NULL)) != -1)
    {
      c = ret;
      switch (c)
        {
          COMMON_ARGS_CASES
        case RECORD_TYPE_ARG:
        case 'd':
          {
            rec2csv_delim = optarg[0];
            break;
          }
        case 't':
          {
            rec2csv_record_type = xstrdup (optarg);
            break;
          }
        case SORT_ARG:
        case 'S':
          {
            if (rec2csv_sort_by_fields)
              {
                recutl_fatal (_("only one list of fields can be specified as a sorting criteria.\n"));
              }

            /* Parse the field name.  */

            if (!rec_fex_check (optarg, REC_FEX_CSV))
              {
                recutl_fatal (_("invalid field name list in -S.\n"));
              }

            rec2csv_sort_by_fields = rec_fex_new (optarg, REC_FEX_CSV);
            if (!rec2csv_sort_by_fields)
              {
                recutl_fatal (_("internal error creating fex.\n"));
              }

            break;
          }
        default:
          {
            exit (EXIT_FAILURE);
          }
        }
    }
}

static void
rec2csv_generate_csv (rec_rset_t rset,
                      rec_fex_t fex)
{
  rec_mset_iterator_t iter;
  rec_fex_elem_t fex_elem;
  rec_record_t record;
  rec_field_t field;
  char *field_name;
  char *tmp;
  size_t i;

  /* Generate the row with headers.  */
  for (i = 0; i < rec_fex_size (fex); i++)
    {
      if (i != 0)
        {
          putc (rec2csv_delim, stdout);
        }

      fex_elem = rec_fex_get (fex, i);
      field_name = xstrdup (rec_fex_elem_field_name (fex_elem));

      /* The header is FNAME or FNAME_N where N is the index starting
         at 1.  Note that we shall remove the trailing ':', if any. */

      if (field_name[strlen(field_name)-1] == ':')
        {
          field_name[strlen(field_name)-1] = '\0';
        }


      if (rec_fex_elem_min (fex_elem) != 0)
        {
          if (asprintf (&tmp, "%s_%d",
                        field_name,
                        rec_fex_elem_min (fex_elem) + 1) == -1)
            recutl_out_of_memory ();
        }
      else
        {
          if (asprintf (&tmp, "%s", field_name) == -1)
            recutl_out_of_memory ();
        }

      csv_fwrite (stdout, tmp, strlen(tmp));
      free (field_name);
      free (tmp);
    }

  putc ('\n', stdout);

  /* Generate the data rows.  */

  iter = rec_mset_iterator (rec_rset_mset (rset));
  while (rec_mset_iterator_next (&iter, MSET_RECORD, (const void**) &record, NULL))
    {
      for (i = 0; i < rec_fex_size (fex); i++)
        {
          if (i != 0)
            {
              putc (rec2csv_delim, stdout);
            }

          fex_elem = rec_fex_get (fex, i);
          field = rec_record_get_field_by_name (record,
                                                rec_fex_elem_field_name (fex_elem),
                                                rec_fex_elem_min (fex_elem));
          if (field)
            {
              csv_fwrite (stdout,
                          rec_field_value (field),
                          strlen (rec_field_value (field)));
            }
        }

      putc ('\n', stdout);
    }

  rec_mset_iterator_free (&iter);
}

static rec_fex_t
rec2csv_determine_fields (rec_rset_t rset)
{
  rec_fex_t fields;
  rec_mset_iterator_t iter_rset;
  rec_mset_iterator_t iter_record;
  rec_record_t record;
  rec_field_t field;
  int field_index;
  
  fields = rec_fex_new (NULL, REC_FEX_SIMPLE);

  iter_rset = rec_mset_iterator (rec_rset_mset (rset));
  while (rec_mset_iterator_next (&iter_rset, MSET_RECORD, (const void **) &record, NULL))
    {
      iter_record = rec_mset_iterator (rec_record_mset (record));
      while (rec_mset_iterator_next (&iter_record, MSET_FIELD, (const void **) &field, NULL))
        {
          field_index = rec_record_get_field_index_by_name (record, field);
          
          if (!rec_fex_member_p (fields,
                                 rec_field_name (field),
                                 field_index, field_index))
            {
              rec_fex_append (fields,
                              rec_field_name (field),
                              field_index, field_index);
            }
        }

      rec_mset_iterator_free (&iter_record);
    }

  rec_mset_iterator_free (&iter_rset);

  return fields;
}

static bool
rec2csv_process_data (rec_db_t db)
{
  bool ret;
  rec_fex_t row_fields;
  size_t i;
  rec_rset_t rset;

  ret = true;

  for (i = 0; i < rec_db_size (db); i++)
    {
      rset = rec_db_get_rset (db, i);
      if (((rec2csv_record_type)
           && rec_rset_type (rset)
           && (strcmp (rec_rset_type (rset),
                       rec2csv_record_type) == 0))
          || (!rec2csv_record_type
              && (!rec_rset_type (rset) ||
                  (rec_db_size (db) == 1))))
        {
          /* Process this record set.  */

          if (!rec_rset_sort (rset, rec2csv_sort_by_fields))
            recutl_out_of_memory ();

          /* Build the fields that will appear in the row. */
          row_fields = rec2csv_determine_fields (rset);
  
          /* Generate the csv data.  */
          rec2csv_generate_csv (rset, row_fields);

          /* Cleanup.  */
          rec_fex_destroy (row_fields);
        }
    }

  return ret;
}

int
main (int argc, char *argv[])
{
  int res;
  rec_db_t db;

  res = 0;

  recutl_init ("rec2csv");

  /* Parse arguments.  */
  rec2csv_parse_args (argc, argv);

  /* Get the input data.  */
  db = recutl_build_db (argc, argv);
  if (!db)
    {
      res = 1;
    }
  else
    /* Process the data.  */
    if (!rec2csv_process_data (db))
      {
        res = 1;
      }

  rec_db_destroy (db);
  
  return res;
}

/* End of rec2csv.c */
